# Introduction
In the sections below, I analyze, clean and glean some data insights from the Chicago Food inspection data set.  A similar analysis can be applied to the other two data sets.

The three data sets 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Load food inspection data from CSV files (Chicago, Las Vegas, New York)
df_ch = pd.read_csv("data/chicago.csv.gz")
df_lv = pd.read_csv("data/las_vegas.csv.gz")
df_ny = pd.read_csv("data/ny.csv.gz")

/home/welshamy/Applications/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


# Exploratory Data Analysis & Data Cleaning
First, let's view the columns the data files have and decide what columns to keep for analytics and model building purposes.

## Chicago

In [3]:
df_ch.columns

Index(['Inspection ID', 'DBA Name', 'AKA Name', 'License #', 'Facility Type',
       'Risk', 'Address', 'City', 'State', 'Zip', 'Inspection Date',
       'Inspection Type', 'Results', 'Violations', 'Latitude', 'Longitude',
       'Location', 'Historical Wards 2003-2015', 'Zip Codes',
       'Community Areas', 'Census Tracts', 'Wards'],
      dtype='object')

In [4]:
df_ch.head()

Inspection ID                         DBA Name  \
0        2288796    CHICAGO HELICOPTER EXPERIENCE   
1        2288787  PROMISELAND LEARNING CENTER INC   
2        2288776                            TOAST   
3        2288798     AFC SUSHI @ JEWEL-OSCO #3501   
4        2288788                 7-ELEVEN #38183B   

                        AKA Name  License #                 Facility Type  \
0  CHICAGO HELICOPTER EXPERIENCE  2658842.0           HELICOPTER TERMINAL   
1    PROMISELAND LEARNING CENTER  2595463.0  Children's Services Facility   
2                          TOAST    46777.0                    Restaurant   
3   AFC SUSHI @ JEWEL-OSCO #3501  2103820.0                 Grocery Store   
4               7-ELEVEN #38183B  2653026.0                 Grocery Store   

              Risk                         Address     City State      Zip  \
0     Risk 3 (Low)              2420 S HALSTED ST   CHICAGO    IL  60608.0   
1    Risk 1 (High)  6201-6205 W MONTROSE AVE BLDG       NaN    IL      NaN   
2    Risk 1 (High)               2046 N DAMEN AVE   CHICAGO    IL  60647.0   
3    Risk 1 (High)             3400 N WESTERN AVE   CHICAGO    IL  60618.0   
4  Risk 2 (Medium)               135 W MADISON ST   CHICAGO    IL  60602.0   

   ...             Results                                         Violations  \
0  ...                Fail  3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...   
1  ...  Pass w/ Conditions  5. PROCEDURES FOR RESPONDING TO VOMITING AND D...   
2  ...  Pass w/ Conditions  3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...   
3  ...  Pass w/ Conditions  3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...   
4  ...  Pass w/ Conditions  3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...   

    Latitude  Longitude                                  Location  \
0  41.848613 -87.646643   (41.84861331486634, -87.64664330114971)   
1  41.960174 -87.781966   (41.96017410031016, -87.78196578638527)   
2  41.919115 -87.677816   (41.91911527657762, -87.67781605299115)   
3  41.943187 -87.688302  (41.943187373182425, -87.68830202070191)   
4  41.881857 -87.631896   (41.88185674363783, -87.63189616155178)   

   Historical Wards 2003-2015 Zip Codes  Community Areas  Census Tracts  Wards  
0                        26.0   14920.0             58.0           59.0   48.0  
1                        25.0   22254.0             15.0          131.0   19.0  
2                        16.0   22535.0             23.0          178.0   40.0  
3                        13.0   21538.0             46.0          110.0   18.0  
4                        22.0   14311.0             38.0           92.0   36.0  

[5 rows x 22 columns]

Since we are building a model to predict food inspection violations and failures, we will only keep features with predictive power.

In [5]:
df_ch = df_ch[["License #", "DBA Name", "Facility Type", 
               "Risk", "Zip", "Inspection Date", 
               "Inspection Type", "Results", "Violations"]]
df_ch

License #                                           DBA Name  \
0       2658842.0                      CHICAGO HELICOPTER EXPERIENCE   
1       2595463.0                    PROMISELAND LEARNING CENTER INC   
2         46777.0                                              TOAST   
3       2103820.0                       AFC SUSHI @ JEWEL-OSCO #3501   
4       2653026.0                                   7-ELEVEN #38183B   
5       2423005.0                   PROMISELAND LEARNING CENTER INC.   
6       2646817.0  HOMEWOOD SUITES BY HILTON CHICAGO DOWNTOWN SOU...   
7       2220948.0                               THE FAT SHALLOT, LLC   
8       2293477.0                                   FIREWATER SALOON   
9       2600677.0                              MCDONALD'S RESTAURANT   
10      2207509.0                         VICTORY CENTER OF GALEWOOD   
11      2646294.0  HOMEWOOD SUITES BY HILTON CHICAGO DOWNTOWN SOU...   
12      2079591.0                              PANADERIA EL ACAMBARO   
13      2665321.0                          GORILLA SUSHI WICKER PARK   
14      2373569.0                                    THE FAT SHALLOT   
15      1948557.0                                      MORE FOR LESS   
16      2579251.0                                ROSAS TAMALES, INC.   
17      2665071.0                                 PUBLIC CLOSEOUT #3   
18      2646295.0  HOMEWOOD SUITES BY HILTON CHICAGO DOWNTOWN SOU...   
19      1495393.0                                       JIMMY MACK'S   
20      2646288.0                                                TBD   
21        34305.0                                         LE BOUCHON   
22      2646290.0                                                TBD   
23      1192909.0                             JEWEL FOOD STORE #3501   
24      2646289.0                                                TBD   
25      2220531.0                                ANTHONY SUPERMARKET   
26      1904672.0                                  SAM FOOD MART INC   
27      1574620.0                             CHILOECHELA RESTAURANT   
28        69016.0                                   GEORGIS CATERING   
29      2271532.0                                TONY'S STEAK & WING   
...           ...                                                ...   
186961  1798522.0                                    POPEYES CHICKEN   
186962  2004254.0                                       Yahs Cuisine   
186963    10940.0                                FAIRPLAY FINER FOOD   
186964    71458.0                                   3811 PIRAMID INC   
186965  2008956.0                       PARK PACKING OF CHICAGO, INC   
186966  2009488.0                                         MAKIA FOOD   
186967  2076658.0                                   THRESHOLD SCHOOL   
186968    18135.0                          MAXWELL STREET DEPOT INC.   
186969  1120626.0                             REGGIO'S PIZZA EXPRESS   
186970  2009114.0                                      WHIPPLE STORE   
186971  2013296.0                                              CITGO   
186972  1991022.0                                            PAKEEZA   
186973  1968898.0                                   KIKKA- KINGSBURY   
186974  1222441.0                                         EAT A PITA   
186975    68917.0                              SULLIVAN'S STEAKHOUSE   
186976  2013344.0                                      DUNKIN DONUTS   
186977    30486.0                                         SEDGWICK'S   
186978  2008633.0                                      WABASH PANTRY   
186979    43513.0                                         EL MILAGRO   
186980    66827.0               SOUTH SIDE FUEL CTR/DUNKIN DOUGHNUTS   
186981    12342.0                                            FERRO'S   
186982  2013259.0                          MONTICELLO FOOD MART, INC   
186983    20575.0                  ST CHRYSOSTOM'S CHURCH DAY SCHOOL   
186984  1992040.0                               

In [6]:
print("Chicago's data after initial feature pruning:\nRows, columns")
df_ch.shape

Chicago's data after initial feature pruning:
Rows, columns


(186991, 9)

Let's start by inspecting the quality of the data and see if the features are properly standardized.
#### `Facility Type`

In [7]:
# Count of food serving facilities by type
counts = df_ch["Facility Type"].value_counts()
print(counts)

Restaurant                                         124229
Grocery Store                                       24048
School                                              11896
Children's Services Facility                         2756
Bakery                                               2718
Daycare (2 - 6 Years)                                2636
Daycare Above and Under 2 Years                      2186
Long Term Care                                       1254
Catering                                             1109
Mobile Food Dispenser                                 846
Liquor                                                833
Daycare Combo 1586                                    780
Mobile Food Preparer                                  558
Golden Diner                                          539
Wholesale                                             530
Hospital                                              504
TAVERN                                                254
Daycare (Under

In [8]:
# Percentage of common type facilities.  Having 100 or more instances of a given type.
counts[counts > 100].sum() / counts.sum()

0.9815134903781256

To simplify our analysis and model development, we will only keep facilities where there are 100 or more instances of its type in our data for a given city.  By doing this, we will lose about 2% of our records.

In [9]:
# Only keep facilities with 100 or more instances of a given type.
df_ch = df_ch.loc[df_ch["Facility Type"].isin(counts[counts > 100].index.values)]

Let's repeat this analyais with other features too.

#### `Inspection Type`

In [10]:
counts = df_ch["Inspection Type"].value_counts()
print(counts)

Canvass                                      93410
License                                      23443
Canvass Re-Inspection                        19173
Complaint                                    17046
License Re-Inspection                         8333
Complaint Re-Inspection                       6981
Short Form Complaint                          6457
Suspected Food Poisoning                       819
Consultation                                   624
Tag Removal                                    594
License-Task Force                             580
Recent Inspection                              294
Task Force Liquor 1475                         242
Suspected Food Poisoning Re-inspection         186
Complaint-Fire                                 155
Short Form Fire-Complaint                      112
Out of Business                                 79
No Entry                                        54
Complaint-Fire Re-inspection                    44
Special Events (Festivals)     

In [11]:
# Percentage of records with a common `Inspection Type` value.
counts[counts > 100].sum() / counts.sum()

0.9976407578646064

In [12]:
df_ch = df_ch.loc[df_ch["Inspection Type"].isin(counts[counts > 100].index.values)]

Let's see the distribution of the `Results` feature.

#### `Results`

In [13]:
df_ch["Results"].value_counts()

Pass                    101271
Fail                     34867
Pass w/ Conditions       23070
Out of Business          12054
No Entry                  5651
Not Ready                 1497
Business Not Located        39
Name: Results, dtype: int64

We only have 7 unique values for the `Results` feature, and they are meaningful, so will keep them all.

#### `Risk`

In [14]:
df_ch["Risk"].value_counts()

Risk 1 (High)      130781
Risk 2 (Medium)     35261
Risk 3 (Low)        12395
All                     6
Name: Risk, dtype: int64

There are 6 records with a value of `All` for the `Risk` feature.  Let's see what their complete records look like.

In [15]:
df_ch[df_ch["Risk"] == "All"]

License #                               DBA Name Facility Type Risk  \
500     2636386.0                       MANHATTAN BAGELS    Restaurant  All   
12259   2608177.0                         111 COFFEE BAR    Restaurant  All   
21256   2506479.0             SAY GRACE RESTAURANT GROUP    Restaurant  All   
39295   2522189.0                       WENDY CITY TACOS    Restaurant  All   
100619  2192690.0  BREWSTONE BEER COMPANY REST & TAPROOM    Restaurant  All   
110398  2204132.0             R N FOOD WINE & SPIRIT INC    Restaurant  All   

            Zip Inspection Date Inspection Type    Results Violations  
500     60657.0      05/01/2019         License  Not Ready        NaN  
12259   60643.0      08/16/2018         License   No Entry        NaN  
21256   60615.0      02/15/2018         License  Not Ready        NaN  
39295   60619.0      04/10/2017         License  Not Ready        NaN  
100619  60654.0      06/24/2014         License       Fail        NaN  
110398  60621.0      12/30/2013         License   No Entry        NaN

We will remove these 6 records as they seem to come from a data entry mistake for this categorical feature.

In [16]:
df_ch = df_ch[df_ch["Risk"] != "All"]

#### `Violations`

In [17]:
df_ch["Violations"].head(10)

1     5. PROCEDURES FOR RESPONDING TO VOMITING AND D...
2     3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...
3     3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...
4     3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...
5                                                   NaN
6                                                   NaN
7     3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL E...
8     25. CONSUMER ADVISORY PROVIDED FOR RAW/UNDERCO...
9     49. NON-FOOD/FOOD CONTACT SURFACES CLEAN - Com...
10    5. PROCEDURES FOR RESPONDING TO VOMITING AND D...
Name: Violations, dtype: object

In [18]:
# Example value of `Violations`
df_ch["Violations"][7]

'3. MANAGEMENT, FOOD EMPLOYEE AND CONDITIONAL EMPLOYEE; KNOWLEDGE, RESPONSIBILITIES AND REPORTING - Comments: NO EMPLOYEE HEALTH POLICY IN PLACE. INSTD TO PROVIDE AND MAINTAIN. PRIORITY FOUNDATION 7-38-010. | 5. PROCEDURES FOR RESPONDING TO VOMITING AND DIARRHEAL EVENTS - Comments: NO PROCEDURE IN PLACE FOR RESPONDING TO VOMITING AND DIARRHEAL EVENTS. INSTD TO PROVIDE PROCEDURE AND ALL SUPPLIES AS NOTED ON PROCEDURE. PRIORITY FOUNDATION 7-38-005.'

The `Violations` feature is a list of the health code violations the facility was cited for.
These violations are number coded.  We are going to extract the code numbers and keep them in an integer list.

In [19]:
# Extract the violations IDs and add them to a new column
a = df_ch["Violations"].str.extractall(r"(^\d+| \d+)\.")
df_ch["Violations IDs"] = a.groupby(level=0)[0].apply(list)
df_ch = df_ch.drop(columns=["Violations"])

#### `Zip`

In [20]:
df_ch["Zip"].isna().value_counts()

False    178365
True         78
Name: Zip, dtype: int64

We have 76 records with missing Zip code values.  We will drop them.

In [21]:
df_ch = df_ch[~df_ch["Zip"].isna()]

#### `Inspection Date`
Convert the inspection date to `datetime` format

In [22]:
df_ch["Inspection Date"] = pd.to_datetime(df_ch["Inspection Date"])

In [23]:
df_ch

License #                                           DBA Name  \
2         46777.0                                              TOAST   
3       2103820.0                       AFC SUSHI @ JEWEL-OSCO #3501   
4       2653026.0                                   7-ELEVEN #38183B   
5       2423005.0                   PROMISELAND LEARNING CENTER INC.   
6       2646817.0  HOMEWOOD SUITES BY HILTON CHICAGO DOWNTOWN SOU...   
7       2220948.0                               THE FAT SHALLOT, LLC   
8       2293477.0                                   FIREWATER SALOON   
9       2600677.0                              MCDONALD'S RESTAURANT   
10      2207509.0                         VICTORY CENTER OF GALEWOOD   
11      2646294.0  HOMEWOOD SUITES BY HILTON CHICAGO DOWNTOWN SOU...   
12      2079591.0                              PANADERIA EL ACAMBARO   
13      2665321.0                          GORILLA SUSHI WICKER PARK   
14      2373569.0                                    THE FAT SHALLOT   
15      1948557.0                                      MORE FOR LESS   
16      2579251.0                                ROSAS TAMALES, INC.   
17      2665071.0                                 PUBLIC CLOSEOUT #3   
18      2646295.0  HOMEWOOD SUITES BY HILTON CHICAGO DOWNTOWN SOU...   
19      1495393.0                                       JIMMY MACK'S   
20      2646288.0                                                TBD   
21        34305.0                                         LE BOUCHON   
22      2646290.0                                                TBD   
23      1192909.0                             JEWEL FOOD STORE #3501   
24      2646289.0                                                TBD   
26      1904672.0                                  SAM FOOD MART INC   
27      1574620.0                             CHILOECHELA RESTAURANT   
28        69016.0                                   GEORGIS CATERING   
29      2271532.0                                TONY'S STEAK & WING   
30      1884826.0                            PANADERIA AYUTLA, CORP.   
31      2246291.0                            KIZIN CREOLE RESTAURANT   
32      2134789.0                                      1ST CHOP SUEY   
...           ...                                                ...   
186959  1998048.0                                 CVS PHARMACY #2935   
186960    18584.0                                    WALGREENS #1417   
186961  1798522.0                                    POPEYES CHICKEN   
186962  2004254.0                                       Yahs Cuisine   
186963    10940.0                                FAIRPLAY FINER FOOD   
186964    71458.0                                   3811 PIRAMID INC   
186965  2008956.0                       PARK PACKING OF CHICAGO, INC   
186967  2076658.0                                   THRESHOLD SCHOOL   
186968    18135.0                          MAXWELL STREET DEPOT INC.   
186969  1120626.0                             REGGIO'S PIZZA EXPRESS   
186970  2009114.0                                      WHIPPLE STORE   
186971  2013296.0                                              CITGO   
186972  1991022.0                                            PAKEEZA   
186973  1968898.0                                   KIKKA- KINGSBURY   
186974  1222441.0                                         EAT A PITA   
186975    68917.0                              SULLIVAN'S STEAKHOUSE   
186976  2013344.0                                      DUNKIN DONUTS   
186977    30486.0                                         SEDGWICK'S   
186978  2008633.0                                      WABASH PANTRY   
186979    43513.0                                         EL MILAGRO   
186981    12342.0                                            FERRO'S   
186982  2013259.0                          MONTICELLO FOOD MART, INC   
186983    20575.0                  ST CHRYSOSTOM'S CHURCH DAY SCHOOL   
186984  1992040.0                               

#### `License #`

In [24]:
df_ch["License #"].value_counts()

0.0          239
1354323.0    190
14616.0      149
1574001.0     69
1974745.0     59
1142451.0     48
1884255.0     47
2083833.0     46
1490035.0     46
1596210.0     45
20481.0       44
60184.0       43
1476553.0     41
1302136.0     40
1094.0        40
1000572.0     40
29151.0       38
2108657.0     38
9154.0        37
25152.0       36
4190.0        36
18234.0       35
1042888.0     35
75883.0       35
1879470.0     35
69637.0       34
80690.0       34
55054.0       34
1448266.0     34
1273271.0     34
            ... 
1738873.0      1
1738871.0      1
1738869.0      1
1738858.0      1
1738855.0      1
27183.0        1
2429101.0      1
2428936.0      1
2428933.0      1
2307885.0      1
2428866.0      1
2428845.0      1
1738916.0      1
2307901.0      1
1739014.0      1
1739023.0      1
2134592.0      1
2252552.0      1
1477265.0      1
2601701.0      1
2601601.0      1
2601607.0      1
1739433.0      1
1921966.0      1
2134584.0      1
36060.0        1
1739571.0      1
1477134.0     

In [25]:
df_ch[df_ch["License #"] == 0.0]

License #                                       DBA Name  \
1302          0.0               ST. GEORGE GREEK ORTHODOX CHURCH   
1313          0.0                    CHURCH OF THE THREE CROSSES   
2086          0.0                                  ALTHEA BY MKC   
2115          0.0                             LITTLE BLACK PEARL   
2502          0.0                                  ALTHEA BY MKC   
4637          0.0                    LUBAVITCH GIRLS HIGH SCHOOL   
5457          0.0                    LUBAVITCH GIRLS HIGH SCHOOL   
6680          0.0                   BELMONT PLACE SENIOR HOUSING   
10918         0.0                                CRAFTS  SERVICE   
11041         0.0                          BIRRIA OCOTLAN MEZCAL   
12216         0.0                            Greater Rock Church   
12386         0.0                            Greater Rock Church   
12474         0.0                               ST. PAULS CHURCH   
15446         0.0                                 SALVATION ARMY   
15513         0.0                   BELMONT PLACE SENIOR HOUSING   
15759         0.0                          ASSUMPTION BVM CHURCH   
17321         0.0                    LUBAVITCH GIRLS HIGH SCHOOL   
17420         0.0               ST. GEORGE GREEK ORTHODOX CHURCH   
18432         0.0                    CHURCH OF THE THREE CROSSES   
18527         0.0                             LITTLE BLACK PEARL   
18991         0.0                         FRIISTYLECHI FENCH FRY   
19062         0.0                                CRAFTS  SERVICE   
19082         0.0                             HAT TRICK CATERING   
24675         0.0                MRS. T'S SOUTHERN FRIED CHICKEN   
25793         0.0                     MA DEARS DOWN HOME KITCHEN   
28407         0.0                MRS. T'S SOUTHERN FRIED CHICKEN   
29272         0.0                MRS. T'S SOUTHERN FRIED CHICKEN   
29729         0.0                          ASADO COFFEE ROASTERS   
30202         0.0                          ASADO COFFEE ROASTERS   
31059         0.0                          SAINT JOSEPH HOSPITAL   
...           ...                                            ...   
159680        0.0                       ST STANISLAUS B&M CHURCH   
160817        0.0                                 FLAV-O-RICIOUS   
161479        0.0                           MATTSONS STEAK HOUSE   
162255        0.0                     St. Vincent De Paul Church   
162474        0.0               ST. GEORGE GREEK ORTHODOX CHURCH   
163694        0.0                  CHICAGO DEPT OF PUBLIC HEALTH   
163851        0.0                    CHURCH OF THE THREE CROSSES   
164915        0.0                                 Anmol Pan Shop   
165172        0.0                                        TAMMY'S   
165392        0.0                                 Anmol Pan Shop   
165717        0.0                    LUBAVITCH GIRLS HIGH SCHOOL   
166019        0.0                    LUBAVITCH GIRLS HIGH SCHOOL   
167042        0.0                                 SALVATION ARMY   
167108        0.0                                 SALVATION ARMY   
167942        0.0                           CLUB DE LA FELICIDAD   
168245        0.0                           CLUB DE LA FELICIDAD   
168835        0.0                                  CHUBBY'S BEEF   
169129        0.0                         MAMA TU'S KORNER STORE   
171648        0.0                 IRISH AMERICAN HERITAGE CENTER   
172634        0.0                              VRJ FOOD MART CO.   
175878        0.0  KENNEDY KING COLLEGE CHILD DEVELOPMENT CENTER   
176245        0.0                          ASSUMPTION BVM CHURCH   
176530        0.0                          NANNY'S WAFFALS HOUSE   
178422        0.0                      DELIVERANCE TEMPLE CHURCH   
178699        0.0          BIBLEWAY APOSTOLIC FAITH CHURCH, INC.   
181321        0.0                         MICHAEL'S FRESH MARKET   
182069        0.0                    CHURCH OF THE THREE CROSSES

We will drop 239 records with `License #` with value `0.0` since it doesn't look like a valid facility license number.  It's probably a data entry error.  The same facility can't have multiple types and different risk classfications.

In [26]:
df_ch = df_ch[df_ch["License #"] != 0.0]

### Areas with high inspection failure rates in 2018
Let's see the zip codes with high rates of inspection failures.  We will filter out zip codes with less than 10 facilities to avoid the small size statistics effect.

In [27]:
# Only keep recrods for 2018
df_ch_2018 = df_ch[df_ch["Inspection Date"].dt.year == 2018]

In [28]:
# Filter out zip codes with less than 10 inspection in that year
counts = df_ch_2018["Zip"].value_counts()
df_ch_2018 = df_ch_2018[df_ch_2018["Zip"].isin(counts[counts > 10].index.values)]

In [29]:
# Calculate failed and all inspections count
df_ch_2018_fail = df_ch_2018[df_ch_2018["Results"] == "Fail"].groupby("Zip")["License #"].count()
df_ch_2018_all = df_ch_2018.groupby("Zip")["License #"].count()

In [30]:
# Inspection failure rates by zip code in descending order
(df_ch_2018_fail / df_ch_2018_all).sort_values(ascending=False)

Zip
60623.0    0.458988
60637.0    0.316456
60621.0    0.310127
60624.0    0.293814
60644.0    0.287129
60827.0    0.285714
60628.0    0.281346
60620.0    0.277778
60659.0    0.276850
60605.0    0.275362
60619.0    0.267123
60629.0    0.262590
60653.0    0.258065
60615.0    0.252874
60647.0    0.239411
60660.0    0.237548
60613.0    0.231121
60616.0    0.228381
60632.0    0.226027
60630.0    0.222222
60639.0    0.219839
60640.0    0.214689
60656.0    0.212121
60603.0    0.211823
60645.0    0.208696
60626.0    0.198653
60609.0    0.198653
60631.0    0.196078
60608.0    0.194805
60651.0    0.191304
60612.0    0.186813
60646.0    0.186813
60641.0    0.183333
60607.0    0.182125
60636.0    0.165563
60625.0    0.159817
60657.0    0.157212
60618.0    0.153285
60666.0    0.152542
60707.0    0.144578
60611.0    0.144348
60649.0    0.142857
60614.0    0.138280
60617.0    0.131410
60610.0    0.127479
60606.0    0.125749
60661.0    0.123377
60654.0    0.115207
60643.0    0.107692
60634.0    0.105

------------

## Las Vegas & New York
We can inspect the columns (features) for the Las Vegas and New York data sets and remove outliers and bad records in a way similar to what we did with the Chicago data set above.

Keep in mind that our three data sets do not share the same feature set.  Even when they overlap on some features, their format and values could be different.  For example, having 5 *vs* 10 digit zip codes, or having different `Inspection Type` values per city.

In [31]:
# Differet data sets may not share the same columns (features)
df_lv.columns

Index(['Serial Number', 'Permit Number', 'Restaurant Name', 'Location Name',
       'Category Name', 'Address', 'City', 'State', 'Zip', 'Current Demerits',
       'Current Grade', 'Date Current', 'Inspection Date', 'Inspection Time',
       'Employee ID', 'Inspection Type', 'Inspection Demerits',
       'Inspection Grade', 'Permit Status', 'Inspection Result', 'Violations',
       'Record Updated', 'Location 1', 'Zip Codes'],
      dtype='object')

In [32]:
df_ny.columns

Index(['CAMIS', 'DBA', 'BORO', 'BUILDING', 'STREET', 'ZIPCODE', 'PHONE',
       'CUISINE DESCRIPTION', 'INSPECTION DATE', 'ACTION', 'VIOLATION CODE',
       'VIOLATION DESCRIPTION', 'CRITICAL FLAG', 'SCORE', 'GRADE',
       'GRADE DATE', 'RECORD DATE', 'INSPECTION TYPE'],
      dtype='object')

In [33]:
# A feature may have different values depending on the city/dataset.
df_lv["Inspection Type"].value_counts()

Routine Inspection               161117
Re-inspection                     17478
Epidemiological Investigation        63
Survey                               41
Name: Inspection Type, dtype: int64

In [34]:
df_ny["INSPECTION TYPE"].value_counts()

Cycle Inspection / Initial Inspection                          223689
Cycle Inspection / Re-inspection                                90110
Pre-permit (Operational) / Initial Inspection                   26850
Pre-permit (Operational) / Re-inspection                        11778
Administrative Miscellaneous / Initial Inspection                6978
Cycle Inspection / Reopening Inspection                          4245
Pre-permit (Non-operational) / Initial Inspection                3457
Smoke-Free Air Act / Initial Inspection                          2757
Administrative Miscellaneous / Re-inspection                     2101
Trans Fat / Initial Inspection                                   1892
Pre-permit (Operational) / Compliance Inspection                 1140
Calorie Posting / Initial Inspection                             1138
Pre-permit (Operational) / Reopening Inspection                   997
Inter-Agency Task Force / Initial Inspection                      992
Cycle Inspection / C

---------

In [35]:
df_ch.head()

License #                                           DBA Name  \
2    46777.0                                              TOAST   
3  2103820.0                       AFC SUSHI @ JEWEL-OSCO #3501   
4  2653026.0                                   7-ELEVEN #38183B   
5  2423005.0                   PROMISELAND LEARNING CENTER INC.   
6  2646817.0  HOMEWOOD SUITES BY HILTON CHICAGO DOWNTOWN SOU...   

                  Facility Type             Risk      Zip Inspection Date  \
2                    Restaurant    Risk 1 (High)  60647.0      2019-05-14   
3                 Grocery Store    Risk 1 (High)  60618.0      2019-05-14   
4                 Grocery Store  Risk 2 (Medium)  60602.0      2019-05-14   
5  Children's Services Facility    Risk 1 (High)  60634.0      2019-05-14   
6                    Restaurant     Risk 3 (Low)  60605.0      2019-05-14   

         Inspection Type             Results  \
2                Canvass  Pass w/ Conditions   
3                Canvass  Pass w/ Conditions   
4                License  Pass w/ Conditions   
5                Canvass     Out of Business   
6  License Re-Inspection                Pass   

                                      Violations IDs  
2  [3,  5,  6,  23,  25,  37,  41,  44,  47,  49,...  
3                                            [3,  5]  
4                                            [3,  5]  
5                                                NaN  
6                                                NaN

In [36]:
df_ch.to_csv("data/chicago_normalized.csv", index=False)

A qeury to create a Hive table for the normalized Chicago data set
```
CREATE EXTERNAL TABLE utilant_food (
    license bigint,
    facility_name string,
    facility_type string,
    risk string,
    zip int,
    inspection_date date,
    inspection_type string,
    results string,
    violations array<string>
)
ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' LINES TERMINATED BY '\n'
LOCATION 's3://utilant-food/normalized';
```

To submit an SQL query to analyze the data, use aws cli.

```
SELECT facility_name, inspection_date, results, violations
FROM utilant_food
WHERE inspection_date>='2019-01-01' and results='Fail';
```